lib/services/local/usagereporter.go

```diff
@@ -0,0 +1,544 @@
+/*
+Copyright 2022 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package local
+
+import (
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"net/http"
+	"time"
+
+	"github.com/bufbuild/connect-go"
+	"github.com/gravitational/trace"
+	"github.com/jonboulle/clockwork"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/sirupsen/logrus"
+	"google.golang.org/protobuf/types/known/timestamppb"
+
+	"github.com/gravitational/teleport"
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/lib/defaults"
+	"github.com/gravitational/teleport/lib/observability/metrics"
+	prehogapi "github.com/gravitational/teleport/lib/prehog/gen/prehog/v1alpha"
+	prehogclient "github.com/gravitational/teleport/lib/prehog/gen/prehog/v1alpha/prehogv1alphaconnect"
+	"github.com/gravitational/teleport/lib/services"
+	"github.com/gravitational/teleport/lib/utils"
+)
+
+const (
+	// usageReporterMinBatchSize determines the size at which a batch is sent
+	// regardless of elapsed time
+	usageReporterMinBatchSize = 20
+
+	// usageReporterMaxBatchSize is the largest batch size that will be sent to
+	// the server; batches larger than this will be split into multiple
+	// requests.
+	usageReporterMaxBatchSize = 100
+
+	// usageReporterMaxBatchAge is the maximum age a batch may reach before
+	// being flushed, regardless of the batch size
+	usageReporterMaxBatchAge = time.Second * 30
+
+	// usageReporterMaxBufferSize is the maximum size to which the event buffer
+	// may grow. Events submitted once this limit is reached will be discarded.
+	// Events that were in the submission queue that fail to submit may also be
+	// discarded when requeued.
+	usageReporterMaxBufferSize = 500
+
+	// usageReporterSubmitDelay is a mandatory delay added to each batch submission
+	// to avoid spamming the prehog instance.
+	usageReporterSubmitDelay = time.Second * 1
```

Comment:
What if we end up generating more than `usageReporterMaxBatchSize` events per second?

Reply:
Once the buffer reaches `usageReporterMaxBufferSize` any new events submitted will be dropped until some events are flushed (and log this somewhat noisily). I figure we can tune these values (delay, batch size, max buffer size) if/when we need to once we have some real world data.